### PR TITLE
ONECOND-1926: Cache auth token

### PR DESCRIPTION
**Pull request by OleksandrOsadchyiGL**

### CODE REVIEW

File: core/src/main/java/com/netflix/conductor/auth/AuthManager.java [hunk 1]
```diff
@@ -209,4 +230,22 @@ public JsonQuery load(@Nonnull String query) throws JsonQueryException {
 		};
 		return CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.HOURS).maximumSize(1000).build(loader);
 	}
+
+	private void cacheAuthData(AuthResponse authResponse) {
+		Map<String, Object> decoded = decode(authResponse.getAccessToken());
+		Object expValue = decoded.get("exp");
```
Comment (lysakaleksey) [inline core/src/main/java/com/netflix/conductor/auth/AuthManager.java:236]: do we need to decode token as expire time already in auth response and it is `expires_in`?

File: core/src/main/java/com/netflix/conductor/auth/AuthManager.java [hunk 2]
```diff
@@ -70,6 +72,8 @@
 	private String authEndpoint;
 	private boolean traceIdEnabled;
 
+	private final AtomicReference<Map<String, Object>> cachedAuthDataRef = new AtomicReference<>();
```
Comment (lysakaleksey) [inline core/src/main/java/com/netflix/conductor/auth/AuthManager.java:75]: Do we need map ? Can't we use tuple.Pair<> from google common lib. I think it is already part of libraries

File: core/src/main/java/com/netflix/conductor/auth/AuthManager.java [hunk 3]
```diff
@@ -97,6 +101,23 @@ public AuthManager(Configuration config) {
 	}
 
 	public AuthResponse authorize(Workflow workflow) throws Exception {
+		synchronized (this) {
```
Comment (lysakaleksey) [inline core/src/main/java/com/netflix/conductor/auth/AuthManager.java:104]: Can we use double synchronized approach but only if `cachedAuthDataRef.get()` has noting in it?

Comment (lysakaleksey) [reply core/src/main/java/com/netflix/conductor/auth/AuthManager.java:104]: Performance shall be better

File: core/src/main/java/com/netflix/conductor/auth/AuthManager.java (hunk 2 above)
Comment (OleksandrOsadchyiGL) [reply core/src/main/java/com/netflix/conductor/auth/AuthManager.java:75]: I decided to make it flexible in order to have a possibility to add another information if needed

File: core/src/main/java/com/netflix/conductor/auth/AuthManager.java (hunk 3 above)
Comment (lysakaleksey) [reply core/src/main/java/com/netflix/conductor/auth/AuthManager.java:104]:
```
Map<String, Object> cachedAuthData = cachedAuthDataRef.get();
if (isNotExpiredAuthData(cachedAuthData)) {
    return (AuthResponse) cachedAuthData.get("authResponse");
}
synchronized (this) {
Map<String, Object> cachedAuthData = cachedAuthDataRef.get(); // read again
// check again
and then only continue 

AuthResponse authResponse = executeAuthorization(workflow);

		if (authResponse.getError() == null && authResponse.getErrorDescription() == null) {
			cacheAuthData(authResponse);
		}
		return authResponse;
}

```

**Pull request closed without merging**